k8s/instance/loop.go

```diff
@@ -0,0 +1,157 @@
+package instance
+
+import (
+	"context"
+	"errors"
+
+	"github.com/deis/steward-framework"
+	"github.com/deis/steward-framework/k8s/data"
+	"github.com/deis/steward-framework/k8s/refs"
+	"k8s.io/client-go/pkg/watch"
+)
+
+var (
+	ErrCancelled     = errors.New("stopped")
+	ErrNotAnInstance = errors.New("not an instance")
+	ErrWatchClosed   = errors.New("watch closed")
+)
+
+// RunLoop starts a blocking control loop that watches and takes action on instance resources
+func RunLoop(
+	ctx context.Context,
+	watchFn WatchInstanceFunc,
+	updateFn UpdateInstanceFunc,
+	getServiceClassFn refs.ServiceClassGetterFunc,
+	getBrokerFn refs.BrokerGetterFunc,
+	lifecycler framework.Lifecycler,
+) error {
+	// TODO: We should watch ALL namespaces; this is temporary.
+	// See https://github.com/deis/steward-framework/issues/29
+	watcher, err := watchFn("default")
+	if err != nil {
+		return err
+	}
+	ch := watcher.ResultChan()
+	defer watcher.Stop()
+	for {
+		select {
+		case <-ctx.Done():
+			return ErrCancelled
+		case evt, open := <-ch:
+			if !open {
+				logger.Errorf("the watch channel was closed")
+				return ErrWatchClosed
+			}
+			switch evt.Type {
+			case watch.Added:
+				if err := handleAddInstance(
+					ctx,
+					lifecycler,
+					updateFn,
+					getServiceClassFn,
+					getBrokerFn,
+					evt,
+				); err != nil {
+					// TODO: try the handler again.
+					// See https://github.com/deis/steward-framework/issues/26
+					logger.Errorf("add instance event handler failed (%s)", err)
+				}
+			case watch.Deleted:
+				if err := handleDeleteInstance(
+					ctx,
+					lifecycler,
+					getServiceClassFn,
+					getBrokerFn,
+					evt,
+				); err != nil {
```

Comment:
FYI, I've created https://github.com/deis/steward-framework/issues/42 since we haven't discussed the global semantics of `Instance` (or other resources) deletion. This is fine as-is in this PR